Upgrade Web Server to use Indy VDR

AlexanderShenshin opened the pull request:
This PR contains `Indy SDK` -> `Indy VDR` migration for VON network Web Server and includes changes adopted from previous work in this direction (see [fork](https://github.com/andrewwhitehead/von-network/tree/indy-vdr)).

AlexanderShenshin commented:
@swcurran, @WadeBarnes ^

WadeBarnes commented on scripts/start_webserver.sh:
```diff
@@ -2,20 +2,13 @@
-		echo von_generate_transactions
-		von_generate_transactions
-	fi
+if [ ! -f "/home/indy/config/genesis.txn" ] && [ -z "${GENESIS_URL}" ] && [ -z "${GENESIS_FILE}" ]; then
+  bash ./scripts/init_genesis.sh
```
The `start_webserver.sh` script, or any of the scripts, do not exist on the web server base image due to how the base images were separated.

AlexanderShenshin replied:
Fixed by reverting base images separation.

WadeBarnes commented on server/anchor.py:
```diff
@@ -175,52 +171,46 @@ class NotReadyException(AnchorException):
-            raise AnchorException("Error opening pool ledger connection") from e
+                LOGGER.info("Finished pool refresh: %s", self._pool.last_status)
+            except VdrError as e:
+                if e.code == VdrErrorCode.POOL_TIMEOUT and attempts < 5:
+                    LOGGER.info("Pool timeout occurred, waiting to retry")
+                    attempts += 1
+                    await asyncio.sleep(10)
+                    continue
```
I like this!  Please make the number of attempts and delay between attempts configurable.

AlexanderShenshin replied:
Done.

WadeBarnes commented on docker/Dockerfile.server:
```diff
@@ -0,0 +1,13 @@
+ADD --chown=indy:indy config ./config
+ADD --chown=indy:indy server ./server
+
+CMD python3 -m server.server
```
The change to the startup `cmd` complicates the genesis file management a bit.

AlexanderShenshin replied:
Removed as we reverted base images separation.

WadeBarnes reviewed (changes requested):
Over all the changes look good.  I have a few general comments here with some specifics below.

### Separation of the Node and Web Server Images
In general I can see why you'd want to separate them, but it does cause some complications.  Additionally it would require updates to the GHA Workflows that build and publish the base image(s).  Having a single base image allows for more consistency across the various scripts and also in how the genesis files are managed and kept in sync, even though it introduces unused artifacts on the image depending on its use case.

### Genesis File Management
How the genesis files are managed has changed with the separation of the images and the addition of the use of the `config/genesis.txn` file.  Adding this additional file complicates genesis file management a bit and actually opens the door for the genesis files used by the nodes and the webserver to inadvertently to get out of sync.

### Broken manage command - generateDid
The generateDid command was inadvertently broken by the separation of the base images.

AlexanderShenshin commented:
Makes sense, these are good points. Thank you for detailed feedback! 
Will be fixed soon.

AlexanderShenshin commented:
@WadeBarnes I pushed PR fixes:
- Reverted separation of Node and Web Server Docker images
- Reverted changes to genesis file management (different paths/scripts), but kept separate `init_genesis` script and updated all startup scripts to use it (to reduce code duplication).
- Added `POOL_CONNECTION_ATTEMPTS` and `POOL_CONNECTION_DELAY` ENV variables for pool connection retries.

AlexanderShenshin commented on scripts/start_synctest.sh:
```diff
@@ -2,18 +2,11 @@
 fi
 
+pip3 install -U pip &&
+pip install --no-cache-dir -r server/requirements.txt
```
Fixed by reverting base images separation.

AlexanderShenshin commented on manage:
```diff
@@ -886,10 +898,12 @@ case "${COMMAND}" in
     ;;
   rebuild)
-      docker build --no-cache --progress plain $(initDockerBuildArgs) -t von-network-base .
+      docker-compose build webserver
```
Not needed as we reverted base images separation.

AlexanderShenshin commented on manage:
```diff
@@ -886,10 +898,12 @@ case "${COMMAND}" in
     ;;
   build)
-      docker build $(initDockerBuildArgs) -t von-network-base .
+      docker-compose build webserver
```
Not needed as we reverted base images separation.

AlexanderShenshin commented on .gitignore:
```diff
@@ -148,13 +148,17 @@ Temporary Items
 config/aml.json
 config/taa.json
 config/.samples.used
+config/genesis.txn
```
Removed as we reverted genesis file management changes.

AlexanderShenshin commented on docker-compose.yml:
```diff
@@ -45,10 +45,8 @@ services:
-      - ./config:/home/indy/config
-      - ./server:/home/indy/server
-      - webserver-cli:/home/indy/.indy-cli
-      - webserver-ledger:/home/indy/ledger
```
Done.

AlexanderShenshin commented on scripts/init_genesis.sh:
```diff
@@ -0,0 +1,25 @@
+    von_generate_transactions
+fi
+
+cp /home/indy/ledger/sandbox/pool_transactions_genesis /home/indy/config/genesis.txn
```
This separate script now used just to reduce code duplication in startup script and uses the same genesis file management approach as before.

WadeBarnes commented on server/anchor.py:
```diff
@@ -63,35 +63,52 @@
-    if isinstance(val, str) and val.isdigit():
-        return True
-    return False
+def format_validator_info(node_data):
+    node_aliases = list(node_data.keys())
+    node_aliases.sort()
+
+    ret = []
+    for node in node_aliases:
+        try:
+            reply = json.loads(node_data[node])
+        except json.JSONDecodeError:
+            data = {"name": node, "error": node_data[node]}  # likely 'timeout'
+        else:
+            if "result" in reply:
+                data = reply["result"]["data"]
+                data["Node_info"]["Name"] = node
+            elif "reason" in reply:
+                data = {"name": node, "error": reply["reason"]}
+            else:
+                data = {"name": node, "error": "unknown error"}
+        ret.append(data)
+    return ret
```
```suggestion
def format_validator_info(node_data):
    node_aliases = list(node_data.keys())
    node_aliases.sort()

    ret = []
    for node in node_aliases:
        try:
            reply = json.loads(node_data[node])
        except json.JSONDecodeError:
            data = {"Node_info": {"Name": node}, "error": node_data[node]}  # likely 'timeout'
        else:
            if "result" in reply:
                data = reply["result"]["data"]
            elif "reason" in reply:
                data = {"Node_info": {"Name": node}, "error": reply["reason"]}
            else:
                data = {"Node_info": {"Name": node}, "error": "unknown error"}
        ret.append(data)
    
    return ret
```

The name of the node needs to be placed into `Node_info.Name` for `/status/text` to work when there is an error result.

Updates are also required to `app.methods.formatValidatorStatus` in `server/static/include/js/index.js` to handle the error results.

Adding:
```
          if (node.error)
            continue
```
to the `for` loop will retain the existing display functionality when nodes encounter an error condition.  In the following example Node 2 was stopped:

![image](https://user-images.githubusercontent.com/10504350/215218630-7e120b83-bc5d-4d98-be89-8d45b8e80069.png)

AlexanderShenshin replied:
Good catch. Done.

WadeBarnes reviewed (changes requested):
Looking good.  That cleaned things up a fair amount.

A couple small issues discovered and detailed below.

WadeBarnes reviewed (commented):
@AlexanderShenshin, Looks good to me.  Do you want to clean-up any of the commit history before I approve and merge, or do you want to leave all of the commits as they are?

AlexanderShenshin commented:
@WadeBarnes I pushed some cleanups for commit names, I think that's it.
Btw, I don't mind if we use squash if needed.
